cmd/earthly/subcmd/build_cmd.go

```diff
@@ -416,6 +312,62 @@ func (a *Build) ActionBuildImp(cliCtx *cli.Context, flagArgs, nonFlagArgs []stri
 		}
 	}
 
+	overridingVars, err := common.CombineVariables(argMap, flagArgs, a.buildArgs.Value())
+	if err != nil {
+		return err
+	}
+
+	err = a.cli.InitFrontend(cliCtx)
+	if err != nil {
+		return errors.Wrapf(err, "could not init frontend")
+	}
+
+	err = a.cli.ConfigureSatellite(cliCtx, cloudClient, gitCommitAuthor, gitConfigEmail)
+	if err != nil {
+		return errors.Wrapf(err, "could not configure satellite")
+	}
+
+	// After configuring frontend and satellites, buildkit address should not be empty.
+	// It should be set to a local container, remote address, or satellite address at this point.
+	if a.cli.Flags().BuildkitdSettings.BuildkitAddress == "" {
+		return errors.New("could not determine buildkit address - is Docker or Podman running?")
+	}
+
+	bkClient, err := buildkitd.NewClient(
+		cliCtx.Context,
+		a.cli.Console(),
+		a.cli.Flags().BuildkitdImage,
+		a.cli.Flags().ContainerName,
+		a.cli.Flags().InstallationName,
+		a.cli.Flags().ContainerFrontend,
+		a.cli.Version(),
+		a.cli.Flags().BuildkitdSettings,
+	)
+	if err != nil {
+		return errors.Wrap(err, "build new buildkitd client")
+	}
+	defer bkClient.Close()
+
+	platr, err := a.platformResolver(cliCtx.Context, bkClient, target)
+	if err != nil {
+		return err
+	}
+
+	skipDB, targetHash, doSkip, err := a.initAutoSkip(cliCtx.Context, target, overridingVars, platr, cloudClient)
```

Comment:
Some variables require the BK client, platform resolver, and Cloud client. I had to reorder some of the initialization here to make those variables available. The auto-skip status now appears under the `Init` section.

```
auto-skip | target +build-args (hash c541ff2117990e1fa9a2f61187b232b3582ce461) has already been run; exiting
```

Reply:
that sounds like some real tight coupling, but unfortunately we need to perform the autoskip detection before creating the client -- since we want it to be as fast as possible. creating the client could be slow for reasons like having to start a new earthly-buildkitd container, or having a satellite wake up.

is it possible to have some sort of lazy client that could be used to skip any buildkit operations? For example, if any buildkit calls are made, it would initially return an error saying it can't be done at this stage. Then once the auto-skip logic has finished (and returned an error saying autoskip detection can't be done), we then connect this lazy client with the real buildkit client.

Reply:
The BK client is used to determine the platform.
https://github.com/earthly/earthly/blob/mh%2Fauto-skip-arg-passing/util/platutil/native.go#L27

We can remove it but we'll lose support for these variables:
https://github.com/earthly/earthly/blob/mh%2Fauto-skip-arg-passing/variables/builtin.go#L37-L41

Perhaps we should assume the platform or just remove support for these variables for the time being.

The lazy client idea could work but it sounds like AS would still depend on a connection if these variables are used.

Reply:
good find regarding the connection to buildkit to determine these.... I think this is just something we can't support and we'll need to simply return a `ErrUnableToDetermineHash` when these are required.

maybe down the road we could allow a way to specify these values locally, but i think we can leave this out for our autoskip-mvp.

Reply:
SGTM. Makes sense.